clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java

```diff
@@ -693,11 +708,33 @@
 
     public static final Schema[] UPDATE_METADATA_REQUEST = new Schema[] {UPDATE_METADATA_REQUEST_V0, UPDATE_METADATA_REQUEST_V1};
     public static final Schema[] UPDATE_METADATA_RESPONSE = new Schema[] {UPDATE_METADATA_RESPONSE_V0, UPDATE_METADATA_RESPONSE_V1};
+    public static final short[] DEPRECATED_UPDATE_METADATA_REQUEST = new short[] {};
+
+    /* Protocol version api */
+    public static final Schema PROTOCOL_VERSION_REQUEST_V0 = new Schema();
+
+    public static final Schema SINGLE_PROTOCOL_VERSION_V0 = new Schema(
+        new Field("api_key", INT16, "Api key of protocol."),
+        new Field("api_name", STRING, "Api name of protocol."),
+        new Field("api_versions", new ArrayOf(INT16), "Supported api versions of protocol in order of preference."),
+        new Field("api_deprecated_versions", new ArrayOf(INT16), "Deprecated api versions of protocol.")
```

Comment:
> The difference here is they would be responsible for writing code that checks/logs it at runtime and running tests with those checks. It will not provide a compile time warning. So regardless, they are expected to test against the new code and look for failures.

True. However, that is true for the whole protocol retrieval thing. Clients, if they want to, will check for supported versions and then choose one. While making that choice, it might be useful to know what versions are deprecated. Again, I do not have a strong feeling in favor and against this. However, after multiple discussions on KIP-35, this is current proposal. If multiple people feel that this is not so much of an requirement, I will be happy to remove this.